File: frontend/bff/src/application/timeline/queries/get-timeline.query.ts
import type { RequestContext } from '../../../cross-cutting/context/request-context.js';
import type { PageRequest } from '../../../cross-cutting/paging/page.js';

/** Filtros da auditoria (todos opcionais). Período em ISO 8601: de (inclusive) até (exclusive). */
export interface TimelineFilters {
  /** Ex.: Policy, Mandate, Order, OrganizationMember. */
  entityType?: string | undefined;
  entityId?: string | undefined;
  /** Created | Updated | Deleted */
  action?: string | undefined;
  authorId?: string | undefined;
  from?: string | undefined;
  to?: string | undefined;
  /** Texto nos valores alterados ou no id da entidade. */
  search?: string | undefined;
}

/** Os mais recentes, até `limit`. */
export interface GetTimelineQuery extends TimelineFilters {
  context: RequestContext;
  limit?: number | undefined;
}

/** Paginado, com o total que atende aos filtros. */
export interface GetTimelinePageQuery extends TimelineFilters {
  context: RequestContext;
  page: PageRequest;
}

File: backend/core-service/src/Fix.Application/UseCases/Timeline/Queries/GetTimelineQuery.cs
using Fix.Application.Abstractions.Authorization;
using Fix.Application.Abstractions.Context;
using Fix.Application.Abstractions.Messaging;
using Fix.Application.Abstractions.Timeline;

namespace Fix.Application.Timeline.Queries;

/// <summary>
/// Auditoria da organização. Filtros opcionais: tipo e id da entidade, ação, autor, período (de inclusive, até
/// exclusive) e texto nos valores alterados. Com <paramref name="Page"/> &gt; 0 vem paginada (com o total);
/// sem página, traz os <paramref name="Limit"/> mais recentes.
/// </summary>
[RequireMembership]
public sealed record GetTimelineQuery(
    Guid UserId,
    Guid OrganizationId,
    string? EntityType,
    Guid? EntityId,
    int Limit,
    string? Action = null,
    Guid? AuthorId = null,
    DateTimeOffset? From = null,
    DateTimeOffset? To = null,
    string? Search = null,
    int Page = 0,
    int PageSize = 0)
    : IQuery<TimelinePage>, IOrganizationRequest;
